spi/src/main/java/org/jboss/pnc/spi/datastore/repositories/api/impl/DefaultSortInfo.java

```diff
@@ -22,36 +22,64 @@
 
 import javax.persistence.criteria.Root;
 import javax.persistence.metamodel.SingularAttribute;
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.List;
 
 public class DefaultSortInfo<T> implements SortInfo<T> {
     private final List<OrderInfo<T>> order;
 
     public DefaultSortInfo(List<OrderInfo<T>> orderInfo) {
-        this.order = List.copyOf(orderInfo);
+        this.order = new ArrayList<>(orderInfo);
     }
 
     public DefaultSortInfo(OrderInfo<T> orderInfo) {
-        this.order = Collections.singletonList(orderInfo);
+        this(Collections.singletonList(orderInfo));
+    }
+
+    public DefaultSortInfo() {
+        this.order = new ArrayList<>();
     }
 
     public static <T> SortInfo<T> asc(SingularAttribute<T, ?> field) {
         DefaultOrderInfo<T> orderInfo = new DefaultOrderInfo<>(
                 OrderInfo.SortingDirection.ASC,
                 (Root<T> root) -> root.get(field));
-        return new DefaultSortInfo(orderInfo);
+        return new DefaultSortInfo<>(orderInfo);
     }
 
     public static <T> SortInfo<T> desc(SingularAttribute<T, ?> field) {
         DefaultOrderInfo<T> orderInfo = new DefaultOrderInfo<>(
                 OrderInfo.SortingDirection.DESC,
                 (Root<T> root) -> root.get(field));
-        return new DefaultSortInfo(orderInfo);
+        return new DefaultSortInfo<>(orderInfo);
+    }
+
+    public static <T> SortInfo<T> descs(SingularAttribute<T, ?>... fields) {
+        return new DefaultSortInfo<T>().appendAll(OrderInfo.SortingDirection.DESC, fields);
+    }
+
+    public static <T> SortInfo<T> ascs(SingularAttribute<T, ?>... fields) {
+        return new DefaultSortInfo<T>().appendAll(OrderInfo.SortingDirection.ASC, fields);
+    }
+
+    private DefaultSortInfo<T> appendAll(OrderInfo.SortingDirection dir, SingularAttribute<T, ?>... fields) {
+        if (fields != null) {
+            for (SingularAttribute<T, ?> attribute : fields) {
+                order.add(new DefaultOrderInfo<>(dir, root -> root.get(attribute)));
+            }
+        }
+
+        return this;
+    }
```

Comment:
public static <T> SortInfo<T> asc(SingularAttribute<T, ?> ... fields) {
        List<DefaultOrderInfo<T>> collect = Arrays.stream(fields)
                .map(f -> new DefaultOrderInfo<T>(OrderInfo.SortingDirection.ASC, (Root<T> root) -> root.get(f)))
                .collect(Collectors.toList());
        return new DefaultSortInfo(collect);
    }
    
    public static <T> SortInfo<T> desc(SingularAttribute<T, ?> ... fields) {
        List<DefaultOrderInfo<T>> collect = Arrays.stream(fields)
                .map(f -> new DefaultOrderInfo<T>(OrderInfo.SortingDirection.DESC, (Root<T> root) -> root.get(f)))
                .collect(Collectors.toList());
        return new DefaultSortInfo(collect);
    }

Reply:
This is a matter of personal preference. I don't see any issues with the current way how it is written.

Reply:
The problem is the mutable protected method appendAll which makes the object mutable.

Reply:
It's mutable just internally inside the class, there is no accessor to the list that is mutable from the outside and all of the public methods available produce new objects. The class is effectively immutable. 

What is different is that multiple calls to the appendAll inside the class itself wont create additional objects.